python-package/lightgbm/sklearn.py

```diff
@@ -812,11 +818,29 @@ def _process_params(self, stage: str) -> Dict[str, Any]:
         params.pop("importance_type", None)
         params.pop("n_estimators", None)
         params.pop("class_weight", None)
+        params.pop("validation_fraction", None)
 
         if isinstance(params["random_state"], np.random.RandomState):
             params["random_state"] = params["random_state"].randint(np.iinfo(np.int32).max)
         elif isinstance(params["random_state"], np.random.Generator):
             params["random_state"] = int(params["random_state"].integers(np.iinfo(np.int32).max))
+
+        params = _choose_param_value(
+            main_param_name="early_stopping_round",
+            params=params,
+            default_value="auto",
+        )
+        if params["early_stopping_round"] == "auto":
+            if hasattr(self, "_n_rows_train") and self._n_rows_train > 10_000:
+                params["early_stopping_round"] = 10
```

Comment:
In https://github.com/microsoft/LightGBM/pull/5808#issuecomment-2506680690, you said:

> *I have set early stopping to be disabled by default*

But this block of code directly contradicts that... it enables early stopping by default if there are more than 10K rows in the training data.

From all the prior discussion on this PR, it seems clear to me that @borchero and @jmoralez both requested you **not** do that, and instead limit the scope of this PR to just providing an API for early stopping that matches how `scikit-learn` estimators do it. I agree with them... it shouldn't be enabled by default as part of this PR.

Please, do the following:

* change the title of the PR (which will become a commit message and release notes item) to `[python-package] add scikit-learn-style API for early stopping`
* remove this logic about `early_stopping_round = "auto"` (and the corresponding changes further down where you moved parameter processing to after input data validation, so it could reference `self._n_rows_train`
* match whatever API `HistGradientBoostingClassifier` / `HistGradientBoostingRegressor` currently have for enabling early stopping

Reply:
Hey @jameslamb, thanks a lot for your review, I highly appreciate that.

You are right with your comment, it was a mistake on my side, I have fixed it in my latest commits. Thank you for spotting it.

In the latest version i had:
- Renamed the PR as you requested.
- Removed the logic about early_stopping_round = "auto" and the corresponding changes where I switched parameters processing and input validation.
- I think the APIs of `HistGradientBoostingClassifier` / `HistGradientBoostingRegressor` and the current implementation are matching except that:
  - `HistGradientBoostingClassifier` / `HistGradientBoostingRegressor` support `early_stopping_round = "auto"` as this is their default parameter value, but as we discussed above we do not want this behaviour.
  - I have tried to support the parameters alias logic of LightGBM while working around the fact that `early_stopping` is an alias for `early_stopping_round(s)` and `n_iter_no_change`. In the current implementation setting `early_stopping_round(s)` to an integer will override the default value for `n_iter_no_change`, but setting `early_stopping` to an integer will not enable early stopping and will not override the default value for  `n_iter_no_change`. I have opted for this implementation to preserve backward compatibility as much as possible. Please let me know if you also agree with it or you have a different opinion, happy to change it.

Reply:
Thanks for that explanation.

> *Removed the logic about early_stopping_round = "auto" and the corresponding changes where I switched parameters processing and input validation.*

Agreed, thank you for removing it.

> *`early_stopping` is an alias for `early_stopping_round(s)` and `n_iter_no_change`

Thanks for being careful about backwards compatibility!

It's unfortunate that we have `early_stopping` as an alias for `early_stopping_rounds` and that `scikit-learn`'s estimators now assign a different meaning to that :/

https://github.com/microsoft/LightGBM/blob/1090a93b39e16f49621aa6824cd09d4390c3678a/include/LightGBM/config.h#L394-L397

https://github.com/scikit-learn/scikit-learn/blob/6cccd99aee3483eb0f7562afdd3179ccccab0b1d/sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py#L1572-L1575

Seems like that has been in `scikit-learn` for a long time (since v0.23):

https://github.com/scikit-learn/scikit-learn/blob/6cccd99aee3483eb0f7562afdd3179ccccab0b1d/sklearn/ensemble/_hist_gradient_boosting/gradient_boosting.py#L1577

I see your implementation of this comment in `LGBMModel._process_params()`:

```python
params = _choose_param_value("early_stopping_round", params, self.n_iter_no_change)
````

I think I see a path to resolving this. It'll be complex, but it would keep LightGBM in compliance with the rules from https://scikit-learn.org/1.5/developers/develop.html#instantiation.

* remove keyword argument `n_iter_no_change` (and property `self.n_iter_no_change`)... but ensure there are unit tests confirming that passing keyword argument `n_iter_no_change` through the scikit-learn estimators correctly sets the number of early stopping rounds
* make keyword argument `early_stopping` have type `Union[bool, int]`
* handle that difference in `LGMBModel._process_params()`, roughly like this:

rewrite this block:

https://github.com/microsoft/LightGBM/blob/1090a93b39e16f49621aa6824cd09d4390c3678a/python-package/lightgbm/sklearn.py#L850-L854

to something like this:

```python
        if stage == "fit":
            # use joblib conventions for negative n_jobs, just like scikit-learn
            # at predict time, this is handled later due to the order of parameter updates
            params = _choose_param_value("num_threads", params, self.n_jobs)
            params["num_threads"] = self._process_n_jobs(params["num_threads"])

            if not isinstance(self.early_stopping, bool) and isinstance(self.early_stopping, int):
                _log_warning(
                    f"Found 'early_stopping={self.early_stopping}' passed through keyword arguments.
                    "Future versions of 'lightgbm' will not allow this, as scikit-learn expects keyword argument "
                    "'early_stopping' to be a boolean indicating whether or not to perform early stopping with "
                    "a randomly-sampled validation set. To set the number of early stopping rounds, and suppress "
                    "this warning, pass early_stopping_rounds={self.early_stopping} instead."
                 )
                 params["early_stopping_round"] = _choose_param_value(
                     main_param_name="early_stopping_round",
                     params=params,
                     default_value=self.early_stopping
                 )
```

And then any other places in the code where you have `self.early_stopping is True` should use `isinstance(self.early_stopping, bool) and self.early_stopping is True`, to avoid this SyntaxWarning:

```python
1 is True
# <stdin>:1: SyntaxWarning: "is" with 'int' literal. Did you mean "=="?
```

I *think* that can work and provide backwards compatibility while allowing us to match scikit-learn's API. And then eventually (after several `lightgbm` releases), that warning could be converted to an error and the type of `early_stopping` could be narrowed to just `bool`.

What do you think?

Reply:
It's a great suggestion! thanks a lot for it, I have implemented it in the latest version.